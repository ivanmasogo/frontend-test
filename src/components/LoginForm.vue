<template>
  <form class="form-signin" @submit.prevent="handleSubmit">
    <img
      class="form-signin__icon mb-4"
      src="@/assets/images/icon-login.svg"
      alt="Login icon"
      width="100"
      height="100"
    />
    <h1 class="form-signin__title mb-3 font-weight-normal">LOGIN</h1>
    <div class="mb-3">
      <div class="form-floating">
        <input
          type="text"
          id="username"
          class="form-control mb-2 border-dark rounded-5"
          placeholder="Username"
          @input="onChange($event, 'username')"
          v-model="user.username"
          aria-describedby="username-error"
        />
        <label for="username"> Username </label>
      </div>
      <ul class="form-signin__errors" v-if="hasUserError" id="username-error">
        <li
          class="form-signin__error-text"
          v-for="(error, index) in errors.username"
          :key="index"
        >
          {{ error }}
        </li>
      </ul>
    </div>
    <div class="form-floating">
      <input
        type="password"
        id="password"
        class="form-control mb-2 border-dark rounded-5"
        placeholder="Password"
        @input="onChange($event, 'password')"
        v-model="user.password"
        aria-describedby="password-error"
      />
      <label for="password"> Password </label>
    </div>
    <ul
      class="form-signin__errors mt-0"
      v-if="hasPasswordError"
      id="password-error"
    >
      <li
        class="form-signin__error-text"
        v-for="(error, index) in errors.password"
        :key="index"
      >
        {{ error }}
      </li>
    </ul>
    <button
      class="form-signin__button btn btn-dark rounded-5 d-flex align-items-center justify-content-center mt-3"
      :disabled="loading"
      type="submit"
    >
      <span v-if="!loading">Sign in</span>

      <!-- Loader -->
      <span
        v-else
        class="loader spinner-border spinner-border-sm"
        role="status"
        aria-hidden="false"
      >
      </span>
    </button>
  </form>
</template>

<script>
import { mapActions } from 'vuex';
import validationMixin from '@/mixins/validationMixin.js';

export default {
  name: 'LoginForm',
  mixins: [validationMixin], // Mixin para la lógica de validación
  data() {
    return {
      loading: false, // Controla el spinner
      errors: {
        // Almacena los errores de validación
        username: null,
        password: null,
      },
      user: {
        username: 'mor_2314', // Usuario por defecto
        password: '83r5^_', // Contraseña por defecto
      },
    };
  },
  computed: {
    hasUserError: function () {
      // True si hay errores en el usuario
      return this.errors.username && this.errors.username.length;
    },
    hasPasswordError: function () {
      // True si hay errores en el password
      return this.errors.password && this.errors.password.length;
    },
  },
  methods: {
    ...mapActions('auth', ['login']),
    /**
     * Maneja el envío del formulario: valida el formulario, intenta iniciar sesión,
     * y proporciona retroalimentación mediante  toasts y redirección de página.
     *
     * @async
     * @returns {Promise<void>}
     */
    async handleSubmit() {
      this.errors = this.validateForm(this.user); // Ejecuta la validación del formulario

      if (this.errors.formIsValid) {
        // Solo se ejecuta si el formulario es válido
        try {
          this.loading = true; // Estado de carga a verdadero

          await this.login(this.user); // Intenta iniciar sesión

          // Mensaje de éxito
          this.$toast.open({
            type: 'success',
            message: 'Login successful.',
          });

          this.$router.push({ name: 'home' }); // Redirige a lahome
        } catch (error) {
          // Mensaje de error
          this.$toast.open({
            type: 'error',
            message: 'Invalid username or password. Please try again.',
          });
        } finally {
          this.loading = false; // Restablece el estado de carga
        }
      }
    },
    /**
     * Maneja el evento de cambio para los campos de entrada: valida el valor de entrada y almacena los errores de validación en 'errors'.
     *
     * @param {Event} e - El evento de entrada activado por el usuario.
     * @param {string} inputName - El nombre del campo de entrada que se está validando.
     * @returns {void}
     */
    onChange(e, inputName) {
      const inputValue = e.target.value; // Obtiene el valor actual de la entrada
      const inputErrors = this.validateField(inputName, inputValue); // Valida el valor de la entrada

      if (inputErrors && inputErrors.length) {
        this.errors[inputName] = inputErrors; // Almacena los errores de validación para el campo
      } else {
        this.errors[inputName] = null; // Elimina los errores si el campo es válido
      }
    },
  },
};
</script>

<style scoped>
.form-signin {
  width: 100%;
  max-width: 500px;
  padding: 15px;
  margin: 0 auto;
}

.form-signin__title {
  font-size: 28px;
  font-weight: bold;
  color: black;
}

.form-signin__errors {
  padding-left: 0 !important;
}

.form-signin__error-text {
  font-size: 14px;
  width: 100%;
  color: red;
  text-align: left;
}

.form-signin__button {
  height: 40px;
  width: 100%;
  height: 30px;
}

.form-signin__button span {
  font-size: 16px;
}
</style>
