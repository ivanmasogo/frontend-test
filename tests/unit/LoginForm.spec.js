import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import LoginForm from '@/components/LoginForm.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

const actions = {
  login: jest.fn(),
};

const store = new Vuex.Store({
  modules: {
    auth: {
      namespaced: true,
      actions,
    },
  },
});

const wrapper = shallowMount(LoginForm, {
  store,
  localVue,
  mocks: {
    $toast: { open: jest.fn() },
    $router: { push: jest.fn() },
  },
});

describe('LoginForm.vue', () => {
  beforeEach(() => {
    jest.clearAllMocks(); // Limpiar todos los mocks antes de cada prueba
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot(); // Compara la salida HTML con el snapshot
  });

  it('shows validation errors when username and password are invalid', async () => {
    const inputUsername = wrapper.find('input#username');
    const inputPassword = wrapper.find('input#password');

    // Entradas vacías
    await inputUsername.setValue('');
    await inputPassword.setValue('');

    // Verificar los errores de validación
    expect(wrapper.vm.errors.username).toBeTruthy();
    expect(wrapper.vm.errors.password).toBeTruthy();
  });

  it('does not show validation errors when username and password are valid', async () => {
    const inputUsername = wrapper.find('input#username');
    const inputPassword = wrapper.find('input#password');

    // Entrada con usuario y contraseña válidos
    await inputUsername.setValue('mor_2314');
    await inputPassword.setValue('83r5^_');

    // Verificar que no haya errores de validación
    expect(wrapper.vm.errors.username).toBeFalsy();
    expect(wrapper.vm.errors.password).toBeFalsy();
  });

  it('calls login action and shows success toast when login is successful', async () => {
    // Entrada válida para el nombre de usuario y la contraseña
    await wrapper.setData({
      user: { username: 'mor_2314', password: '83r5^_' },
    });

    // Simular el envío del formulario
    await wrapper.vm.handleSubmit();

    // Esperar que la acción de login sea llamada con el usuario y contraseña correctos
    expect(actions.login).toHaveBeenCalledWith(expect.anything(), {
      username: 'mor_2314',
      password: '83r5^_',
    });

    // Esperar que se muestre el toast de éxito
    expect(wrapper.vm.$toast.open).toHaveBeenCalledWith({
      type: 'success',
      message: 'Login successful.',
    });

    // Esperar que el router redirija a la página de inicio
    expect(wrapper.vm.$router.push).toHaveBeenCalledWith({ name: 'home' });
  });

  it('shows error toast when login fails', async () => {
    // Simular un fallo en la acción de login
    actions.login.mockRejectedValue(
      new Error('An error occurred. Please try again later.')
    );

    // Simular entrada inválida de usuario
    await wrapper.setData({
      user: { username: 'mor_2314', password: '83r5^_1' },
    });

    // Simular el envío del formulario
    await wrapper.vm.handleSubmit();

    // Esperar que se muestre el toast de error
    expect(wrapper.vm.$toast.open).toHaveBeenCalledWith({
      type: 'error',
      message: 'Invalid username or password. Please try again.',
    });
  });
});
