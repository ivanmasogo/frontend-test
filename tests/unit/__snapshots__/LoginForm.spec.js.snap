// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm.vue matches the snapshot 1`] = `
<form class="form-signin"><img src="@/assets/images/icon-login.svg" alt="Login icon" width="100" height="100" class="form-signin__icon mb-4">
  <h1 class="form-signin__title mb-3 font-weight-normal">LOGIN</h1>
  <div class="mb-3">
    <div class="form-floating"><input type="text" id="username" placeholder="Username" aria-describedby="username-error" class="form-control mb-2 border-dark rounded-5"> <label for="username"> Username </label></div>
    <!---->
  </div>
  <div class="form-floating"><input type="password" id="password" placeholder="Password" aria-describedby="password-error" class="form-control mb-2 border-dark rounded-5"> <label for="password"> Password </label></div>
  <!----> <button type="submit" class="form-signin__button btn btn-dark rounded-5 d-flex align-items-center justify-content-center mt-3"><span>Sign in</span></button>
</form>
`;
